### ISSUE TITLE
Psi staircase not working?

### ISSUE THREAD

**Issue by hoechenberger** [psychopy/psychopy#1013]
I cannot seem to get the Psi adaptive staircase to work. It always sticks to the lowest possible intensity value. This issue was also [reported](https://groups.google.com/d/msg/psychopy-users/frva_zDMBEM/pw8Ff8O6T2gJ) by another user on the mailing list.

Example code:
```python
from __future__ import print_function
from psychopy.data import PsiHandler

if __name__ == '__main__':
    nTrials = 10
    intensRange = [0.01, 100]
    alphaRange = [0.01, 100]
    betaRange = [0.01, 10]
    delta = 0.01
    intensPrecision, alphaPrecision, betaPrecision = 10000, 100, 100
    stepType = 'lin'
    TwoAFC = False
    prior = None

    stairs = PsiHandler(
        nTrials, intensRange, alphaRange, betaRange, delta,
        intensPrecision, alphaPrecision, betaPrecision, TwoAFC=TwoAFC,
        stepType=stepType, prior=prior
    )

    responses = [1, 1, 0, 0, 1, 1, 0, 0, 1, 1]

    for trialN, intensityN in enumerate(stairs):
        stairs.addResponse(responses[trialN])
        print(intensityN)
```
Output:
```
0.01
0.01
0.01
0.01
0.01
0.01
0.01
0.01
0.01
0.01
```
@JosephGlavan, maybe you have an idea? :)

**Comment by peircej** [comment]
Yes, I recall the post on the users forum and haven't had a chance to look at it. (I don't know much about psi method)

**Comment by hoechenberger** [comment]
I read the paper yesterday (http://jov.arvojournals.org/article.aspx?articleid=2121598), but didn't have time yet to verify the whole algorithm implemented in `contrib.psi`. I'm afraid something fishy is going on in there, and not in `data.PsiHandler`.

**Comment by JosephGlavan** [comment]
Hey all. Sorry it took me so long to get back to you; I've been out of town
for a conference. The problem is with how PsiHandler is being initialized.
The three "precision" keywords (intensPrecision, alphaPrecision,
betaPrecision) all specify the step sizes of the array to search over, not
the number of step (with one exception, more on that in a second). Because
10,000 does not fit into the range [0.01, 100], PsiHandler is being
initialized with the array [0.01]. This is why it continues to print 0.01
no matter what you do; it's the only intensity it knows.

The exception to this is when the stepType is set to 'log'. Then
intensPrecision specifies the number of steps, as was intended here.
alphaPrecision and betaPrecision would still specify the size of the steps
though. Admittedly this might be confusing for the user, and if it's
generally agreed by everyone that it should be changed I can. The logspace
functionality was added in at the end which is why it sort of seems like an
afterthought. I implemented it the way I did because the step size is what
I thought is important to the experimenter, whereas number of steps is
really only useful for figuring out if you have enough RAM to store the
PsiHandler object. Unfortunately step size is not meaningful for log space.



Now the issue brought up in the mailing list is different. It's a bug (I
feel terrible to admit) I found a couple months ago specific to the 2AFC
part of the code. I have a fix for it; I just never uploaded it because
I've been busy with other things and it seemed like no one had found the
bug yet... Look for a pull request with the fix in about an hour. In the
future I'll try not to be a scumbag and fix things as I find them. Do you
know if there's any way I can get psi-specific notifications from the
mailing list or do I have to just run a search on it every so often?

-Joe

**Comment by JosephGlavan** [comment]
Pull requests submitted.

**Comment by hoechenberger** [comment]
Thanks Joe! It appears to be working now:
```python
from __future__ import print_function
from psychopy.data import PsiHandler

if __name__ == '__main__':
    nTrials = 10
    intensRange = [0.01, 100]
    alphaRange = [0.01, 10]
    betaRange = [0.01, 10]
    delta = 0.01
    intensPrecision, alphaPrecision, betaPrecision = 1, 0.1, 0.1
    stepType = 'lin'
    TwoAFC = False
    prior = None

    stairs = PsiHandler(
        nTrials, intensRange, alphaRange, betaRange, delta,
        intensPrecision, alphaPrecision, betaPrecision, TwoAFC=TwoAFC,
        stepType=stepType, prior=prior
    )

    responses = [1, 1, 0, 0, 1, 1, 0, 0, 1, 1]

    for trial, intensity in enumerate(stairs):
        stairs.addResponse(responses[trial])
        print(intensity)
```
Output:
```python
4.71
3.51
2.54
4.76
6.24
5.97
2.16
6.25
7.53
1.47
```
The irregular meaning of the step sizes is very confusing indeed, though. Would be great if that could be clarified at least in the documentation, or, better yet, the code be improved in this regard :)

I think there is no way to automatically poke you when Psi issues are brought up on the mailing list, but I think we can ensure to mention you (via @JosephGlavan) in Psi Github issues.

Thanks for your help! :)

**Comment by hoechenberger** [comment]
Just for reference, the regarding pull requests are #1024 and #1025.

**Comment by hoechenberger** [comment]
This is fixed via #1024 and #1025.
